idea-plugin/src/main/kotlin/arrow/meta/ide/dsl/ui/popups/PopupSyntax.kt

```diff
@@ -0,0 +1,150 @@
+import com.intellij.openapi.ui.popup.TreePopup
+import com.intellij.openapi.ui.popup.TreePopupStep
+import com.intellij.openapi.ui.popup.util.BaseListPopupStep
+import java.awt.Color
+import javax.swing.Icon
+import javax.swing.JComponent
+import javax.swing.event.HyperlinkEvent
+
+/**
+ * Popup's can be created at any time.
+ * Either in [AnActionSyntax], [InspectionSyntax] or [IntentionExtensionProviderSyntax] and many other places.
+ * Once the PopUp is created the user needs to call a `show()` method like [com.intellij.openapi.ui.popup.JBPopup.showUnderneathOf], this is an effectfull operation.
+ */
+interface PopupSyntax {
+  /**
+   * ```kotlin:ank:playground
+   * import arrow.meta.Plugin
+   * import arrow.meta.ide.IdeMetaPlugin
+   * import arrow.meta.ide.dsl.ui.popups.IdeListPopupItem
+   * import arrow.meta.ide.resources.ArrowIcons
+   * import arrow.meta.invoke
+   * import com.intellij.openapi.actionSystem.PlatformDataKeys
+   *
+   * val IdeMetaPlugin.showListPlugin: Plugin
+   *  get() = "Action to create ListPopUp" {
+   *    meta(
+   *      addAnAction(
+   *        "Unique",
+   *        anAction(
+   *          title = "Show ListPopUp",
+   *          description = "Shows one item of a ListPopUp",
+   *          icon = ArrowIcons.ICON4,
+   *          actionPerformed = { e ->
+   *            PlatformDataKeys.CONTEXT_COMPONENT.getData(e.dataContext)?.run {
+   *              listPopUp(
+   *                title = "TopLevelTitle",
+   *                sameIcon = ArrowIcons.ICON3, // every PopUp item will have this Icon
+   *                popUps = listOf(
+   *                  IdeListPopupItem(
+   *                    this,
+   *                    text = { "Teach your users about this Component: $this" }
+   *                  )
+   *                )
+   *              ).show(this)
+   *            }
+   *          }
+   *        )
+   *      )
+   *    )
+   *  }
+   * ```
+   * @param A does not necessarily have to be the same as the element you pass to a `show` Function
+   * @param canceled allows post-processing
+   * @param sameIcon can be specified to unify all ListPopUp items
+   * @see IdeListPopupItem
+   */
+  fun <A> IdeMetaPlugin.listPopUp(
+    popUps: List<IdeListPopupItem<A>>,
+    title: String? = null,
+    sameIcon: Icon? = null,
+    canceled: () -> Unit = {}
+  ): ListPopup =
+    popUp {
+      popUps.run {
+        createListPopup(object : BaseListPopupStep<A>(title, map { it.element }, sameIcon) {
+          override fun isSelectable(value: A): Boolean =
+            this@run.firstOrNull { it == value }?.isSelectable ?: false
+
+          override fun hasSubstep(selectedValue: A): Boolean =
+            this@run.firstOrNull { it == selectedValue }?.hasSubStep ?: false
+
+          override fun onChosen(selectedValue: A, finalChoice: Boolean): PopupStep<*> =
+            this@run.firstOrNull { it == selectedValue }?.let { a -> a.onChosen(selectedValue, finalChoice) }
```

Comment:
Is `a` meant to also be `a: A`?  perhaps this can be made a little clearer, like `selectedValueType`